### PR TITLE
[i.MX CAAM] SGT + Hash Fix

### PR DESCRIPTION
**Pull request by clementfaure**
Hi everyone,

This PR introduces the scatter gather for CAAM driver. CAAM uses DMA to load/store data in memory. In case of a user buffer, if the buffer is crossing multiple Small Page, a CAAM Scatter Gather Table needs to be created to rebuild the physical memory chunks used by the User virtual buffer.

This PR also add the use of SGT for the Hash driver.

Thanks!

### CODE REVIEW

File: core/arch/arm/plat-imx/crypto_conf.mk [hunk 1]
```diff
@@ -8,14 +8,15 @@
...
 # DBG_TRACE_MEM  BIT32(2)  // Memory utility trace
-# DBG_TRACE_PWR  BIT32(3)  // Power trace
-# DBG_TRACE_JR   BIT32(4)  // Job Ring trace
-# DBG_DESC_JR    BIT32(5)  // Job Ring dump descriptor
-# DBG_TRACE_RNG  BIT32(6)  // RNG trace
-# DBG_DESC_RNG   BIT32(7)  // RNG dump descriptor
-# DBG_TRACE_HASH BIT32(8)  // Hash trace
-# DBG_DESC_HASH  BIT32(9)  // Hash dump descriptor
-# DBG_BUF_HASH   BIT32(10) // Hash dump Buffer
+# DBG_TRACE_SGT  BIT32(3)  // Scatter Gather trace
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/arch/arm/plat-imx/crypto_conf.mk]: I agree, I will fix it.

File: core/drivers/crypto/caam/include/caam_utils_sgt.h [hunk 2]
```diff
@@ -16,6 +16,40 @@
...
+void caam_sgt_set_entry(struct caamsgt *sgt, vaddr_t paddr, size_t len,
+			unsigned int offset, bool final_e, bool ext_e);
+
+#define CAAM_SGT_ENTRY(sgt, paddr, len)                                        \
+	caam_sgt_set_entry(sgt, paddr, len, 0, false, false)
+#define CAAM_SGT_ENTRY_FINAL(sgt, paddr, len)                                  \
+	caam_sgt_set_entry(sgt, paddr, len, 0, true, false)
+#define CAAM_SGT_ENTRY_EXT(sgt, paddr, len)                                    \
+	caam_sgt_set_entry(sgt, paddr, len, 0, false, true)
+/*
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/include/caam_utils_sgt.h]: Ok

File: core/drivers/crypto/caam/include/caam_utils_sgt.h [hunk 3]
```diff
@@ -16,6 +16,40 @@
...
+ * Build a SGT object with @block and @data buffer.
+ * If @block is defined, create a SGT with block buffer as first SGT entry
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/include/caam_utils_sgt.h]: Ok!

File: core/drivers/crypto/caam/utils/utils_sgt.c [hunk 4]
```diff
@@ -5,18 +5,174 @@
...
+}
+
+static void caam_sgt_fill_table(struct caambuf *buf, struct caamsgtbuf *sgt,
+				int start_idx, int nb_pa)
+{
+	int idx = 0;
+
+	SGT_TRACE("Create %d SGT entries", nb_pa);
+
+	for (; idx < nb_pa; idx++) {
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_sgt.c]: Ok

File: core/drivers/crypto/caam/utils/utils_sgt.c [hunk 5]
```diff
@@ -5,18 +5,174 @@
...
+	struct caambuf *pabufs = NULL;
+
+	/* Get the number of physical areas used by the input buffer @data */
+	nb_pa_area = caam_mem_get_pa_area(data, &pabufs);
+	if (nb_pa_area == -1)
+		return CAAM_FAILURE;
+
+	/*
+	 * If the block buffer is present, we need a SGT object
+	 * with a minimum of 2 entries. In plus, if the data is mapped
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_sgt.c]: Ok

File: core/drivers/crypto/caam/utils/utils_sgt.c [hunk 6]
```diff
@@ -5,18 +5,174 @@
...
+	 * on non-contiguous physical areas, we need a SGT object with
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_sgt.c]: Ok

File: core/drivers/crypto/caam/include/caam_desc_helper.h [hunk 7]
```diff
@@ -143,6 +143,15 @@ static inline void dump_desc(uint32_t *desc)
 	(CMD_FIFO_LOAD_TYPE | CMD_CLASS(cla) | CMD_SGT |                       \
 	 FIFO_LOAD_INPUT(dst) | FIFO_LOAD_ACTION(act) | FIFO_LOAD_LENGTH(len))
 
+/*
+ * FIFO Load to register dst class cla with action act.
+ * Pointer is a Scatter/Gatter Table
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/include/caam_desc_helper.h]: I will fix it

File: core/drivers/crypto/caam/utils/utils_sgt.c [hunk 8]
```diff
@@ -5,18 +5,174 @@
...
+	 * the number of physical area + one entry for the block buffer.
+	 *
+	 * If the block buffer is not present and data is mapped
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_sgt.c]: Ok

**Comment by clementfaure** [comment]
@etienne-lms @jenswi-linaro
You guys have been very helpful for the first CAAM pull-request.
Could you spend some time reviewing this one?

**Comment by jenswi-linaro** [comment]
I assume that this doesn't work with paging of TAs enabled. For that we'd need something like `get_usage_pages()` in the Linux kernel. It's not in the scope of this PR, but long term we'll need to do something about it.

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by clementfaure** [comment]
Indeed, it won't work with TA paging enabled for now ...
Patch review squashed and tags applied

**Comment by clementfaure** [comment]
Any more comments on this PR?

**Comment by jforissier** [comment]
@clementfaure no comment from me at least, this looks good to be merged. @etienne-lms I think all your comments have been addressed, do you want to provide a tag?

**Pull request merged**